contrib/babelfishpg_tsql/src/pltsql_ruleutils.c

```diff
@@ -377,6 +379,80 @@ tsql_get_constraintdef(PG_FUNCTION_ARGS)
 	PG_RETURN_TEXT_P(string_to_text(res));
 }
 
+PG_FUNCTION_INFO_V1(tsql_get_expr);
+
+/* ----------
+ * tsql_get_expr          - Decompile an expression tree
+ *
+ * Input: an expression tree in nodeToString form, and a relation OID
+ *
+ * Output: reverse-listed expression
+ *
+ * Currently, the expression can only refer to a single relation, namely
+ * the one specified by the second parameter.  This is sufficient for
+ * partial indexes, column default expressions, etc.  We also support
+ * Var-free expressions, for which the OID can be InvalidOid.
+ * ----------
+ */
+
+Datum
+tsql_get_expr(PG_FUNCTION_ARGS)
+{
+    text       *expr = PG_GETARG_TEXT_PP(0);
+    Oid         relid = PG_GETARG_OID(1);
+    int         prettyFlags;
+    char       *relname;
+
+    prettyFlags = PRETTYFLAG_INDENT;
+
+    if (OidIsValid(relid))
+    {
+        /* Get the name for the relation */
+        relname = get_rel_name(relid);
+
+        /*
+         * If the OID isn't actually valid, don't throw an error, just return
+         * NULL.  This is a bit questionable, but it's what we've done
+         * historically, and it can help avoid unwanted failures when
+         * examining catalog entries for just-deleted relations.
+         */
+        if (relname == NULL)
+            PG_RETURN_NULL();
+    }
+    else
+        relname = NULL;
+
+    PG_RETURN_TEXT_P(tsql_get_expr_worker(expr, relid, relname, prettyFlags));
+}
+
+static text *
+tsql_get_expr_worker(text *expr, Oid relid, const char *relname, int prettyFlags)
+{
+    Node       *node;
+    List       *context;
+    char       *exprstr;
+    char       *str;
+
+    /* Convert input TEXT object to C string */
+    exprstr = text_to_cstring(expr);
+
+    /* Convert expression to node tree */
+    node = (Node *) stringToNode(exprstr);
+
+    pfree(exprstr);
+
+    /* Prepare deparse context if needed */
+    if (OidIsValid(relid))
+        context = deparse_context_for(relname, relid);
```

Comment:
can this function handle relname being NULL?

Reply:
you mean to say deparse_context_for can handle relname being null? please validate with test case

Reply:
I actually mean that deparse_context_for is only called when OID is valid and if OID is valid then relname will surely be not NULL because if it is then it is handled in tsql_get_expr function. I have added testcase though.

Reply:
Did the required changes.